#40758: Tweaks to the bootstrap API.

thebeeland opened the pull request:
Allows for the following:

* Easy access to the resolved pipeline config’s descriptor.
* Sorting/filtering of pre-queried PipelineConfiguration entities.

jfboismenu commented on python/tank/bootstrap/manager.py:
```diff
@@ -546,37 +560,30 @@ def get_pipeline_configurations(self, project):
-        # Only return id, type and code fields.
-        pcs = []
-        for pc in resolver.find_matching_pipeline_configurations(
+        return resolver.find_matching_pipeline_configurations(
```
I should have written a unit test. There is code out there expecting name and not code.

jfboismenu replied:
And by out there, I mean [here](https://github.com/shotgunsoftware/tk-desktop/blob/develop/zero_config/python/tk_desktop/desktop_window.py#L671-L674)

thebeeland replied:
Ah, good point, and I prefer "name" anyway. Really, the only thing I needed was the entity type included, so I'll just add that to the old structure. I preferred to not hardcode it in my caching script.

thebeeland replied:
What I wrote above is a lie. I now have no recollection as to why I made this specific change. Reverting it now.

jfboismenu commented on python/tank/bootstrap/manager.py:
```diff
@@ -516,7 +516,21 @@ def prepare_engine(self, engine_name, entity):
-    def get_pipeline_configurations(self, project):
+    def get_resolved_pipeline_configuration_descriptor(self, project_id=None):
+        """
+        Resolves a pipeline configuration and returns its associated descriptor object.
```
Could you provide a bit more context for that change? Why do you need this? Maybe a link to your work in progress code for the websocket code? Thanks.

thebeeland replied:
The config descriptor's URI is used as part of the unique identifier for a set of engine commands in the sqlite cache that's being build by the websocket server's RPC api. As such, I need a descriptor.

jfboismenu replied:
Why is a descriptor required? Would just the pipeline configuration is be sufficient? After all, one is interchangeable for the other.

Unless you are expecting people to use the same descriptor with multiple projects?

jfboismenu replied:
Ah... yes. I think I got it now. Is that because we also have to deal with the fallback, which doesn't have an id and could be any value?

thebeeland replied:
Correct. We're in a situation where something like tk-config-basic could (and likely will) be used by multiple projects. In that situation we can reuse the already-cached data, even if it wasn't cached by the current project.

thebeeland commented:
Hey @manneohrstrom , what's your take on sort/filter of already-queried PC entities? I'm fine with making a new method instead of piggy-backing on the existing getter if that sounds better to you and JF.

manneohrstrom commented on python/tank/bootstrap/cached_configuration.py:
```diff
@@ -88,6 +88,13 @@ def __repr__(self):
         )
 
+    @property
+    def descriptor(self):
```
what about other `Configuration` subclasses? Seems we should display this in some form on the baseclass to provide a default for all deriving classes?

manneohrstrom commented on python/tank/bootstrap/manager.py:
```diff
@@ -516,7 +516,21 @@ def prepare_engine(self, engine_name, entity):
         return path
 
-    def get_pipeline_configurations(self, project):
+    def get_resolved_pipeline_configuration_descriptor(self, project_id=None):
```
this name is very long. Could we come up with something shorter? `resolve_descriptor()`?

manneohrstrom replied:
could this be symmetrical with `def get_pipeline_configurations(self, project):` and take a project dict (and no `None` default value)?

thebeeland replied:
Yep! I agree with both comments.

manneohrstrom commented on python/tank/bootstrap/manager.py:
```diff
@@ -516,7 +516,21 @@ def prepare_engine(self, engine_name, entity):
+        Resolves a pipeline configuration and returns its associated descriptor object.
+
+        :param int project_id: The id of the project to resolve a configuration for, or None.
+        """
```
missing returns docstring

manneohrstrom commented on python/tank/bootstrap/manager.py:
```diff
@@ -546,15 +560,17 @@ def get_pipeline_configurations(self, project):
             If ``None``, this will enumerate the pipeline configurations
             for the site configuration.
         :type project: Dictionary with keys ``type`` and ``id``.
+        :param list pc_entities: A list of PipelineConfiguration entity dictionaries. This
```
Just a subjective opinion, but maybe this could be named something like `external_data` instead? To highlight that instead if having the method fetch the data, it just operates directly on this external data that is passed? Feeel free to ignore this if you disagree! :)

manneohrstrom commented:
@thebeeland @jfboismenu specifically regarding "what's your take on sort/filter of already-queried PC entities? I'm fine with making a new method instead of piggy-backing on the existing getter if that sounds better to you and JF." -- i don't have a strong opinion of this either, but in my head i was expecting what jeff did, just add a parameter which lets you pass the data directly to the method instead of having the method querying for it. I suggested calling the param `external_data` in my CR comments above - maybe that makes it clearer?

To me it feels like the same method since the only difference is that in one case we pass in external data to be processed and in the other case we ask the method to retrieve data.

jfboismenu commented:
I think we're much better served by having two methods, each with a clear role, than having one that can operate in two different modes and that is harder to document.

thebeeland commented:
I split it into two methods, as JF feels strongly about it and Manne and I do not.

Added tests, which themselves exposed bugs in my initial implementation. Those have not been sorted out, as well.

jfboismenu commented on python/tank/bootstrap/resolver.py:
```diff
@@ -422,13 +422,30 @@ def find_matching_pipeline_configurations(self, pipeline_config_name, current_lo
         """
-        pcs = self._get_pipeline_configurations_for_project(pipeline_config_name, current_login, sg_connection)
+        # Filter out anything not from the current project.
+        if external_data:
```
Won't this filter out site configurations that apply to the current project?
